File: week01/example/ltc150_evaluate_reverse_polish_notation.py
from typing import List


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if token in ["+", "-", "*", "/"]:
                y = stack.pop()
                x = stack.pop()
                z = self.calc(x, y, token)
                stack.append(z)
            else:
                stack.append(int(token))
        return stack[-1]

    def calc(self, x, y, op):
        if op == "+": return x + y
        if op == "-": return x - y
        if op == "*": return x * y
        if op == "/": return int(x / y) # 需要注意 python 中负数除法的表现与题目不一致
        return 0


if __name__ == '__main__':
    tokens = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]
    expect_res = 22
    out = Solution().evalRPN(tokens)
    assert expect_res == out

File: week01/example/ltc150_evaluate_reverse_polish_notation.cpp
//
// Created by Opt_c on 2022/4/21.
//
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int evalRPN(vector<string> &tokens) {
        for (string &token: tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                int y = s.top();
                s.pop();
                int x = s.top();
                s.pop();
                int z = calc(x, y, token);
                s.push(z);
            } else {
                s.push(atoi(token.c_str()));
            }

        }
        return s.top();

    }

private:
    stack<int> s;

    int calc(int x, int y, string &op) {
        if (op == "+") return x + y;
        if (op == "-") return x - y;
        if (op == "*") return x * y;
        if (op == "/") return x / y;
        return 0;
    }
};